Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyFilePathCtrl.cpp

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "PropertyFilePathCtrl.h"
+
+AZ_PUSH_DISABLE_WARNING(4244 4251, "-Wunknown-warning-option")
+#include <QDir>
+#include <QFileDialog>
+#include <QHBoxLayout>
+#include <QLineEdit>
+#include <QMessageBox>
+#include <QToolButton>
+AZ_POP_DISABLE_WARNING
+
+#include <AzCore/Utils/Utils.h>
+
+#include <AzQtComponents/Components/Widgets/BrowseEdit.h>
+#include <AzQtComponents/Components/Widgets/FileDialog.h>
+
+#include <AzToolsFramework/API/EditorAssetSystemAPI.h>
+#include <AzToolsFramework/UI/UICore/WidgetHelpers.h>
+
+namespace AzToolsFramework
+{
+    PropertyFilePathCtrl::PropertyFilePathCtrl(QWidget* parent)
+        : QWidget(parent)
+    {
+        QHBoxLayout* layout = new QHBoxLayout(this);
+
+        m_browseEdit = new AzQtComponents::BrowseEdit(this);
+        m_browseEdit->lineEdit()->setFocusPolicy(Qt::StrongFocus);
+        m_browseEdit->setLineEditReadOnly(true);
+        m_browseEdit->setClearButtonEnabled(true);
+        m_browseEdit->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
+        QObject::connect(m_browseEdit, &AzQtComponents::BrowseEdit::attachedButtonTriggered, this, &PropertyFilePathCtrl::OnOpenButtonClicked);
+
+        QToolButton* clearButton = AzQtComponents::LineEdit::getClearButton(m_browseEdit->lineEdit());
+        AZ_Assert(clearButton, "Clear button missing from BrowseEdit");
+        QObject::connect(clearButton, &QToolButton::clicked, this, &PropertyFilePathCtrl::OnClearButtonClicked);
+
+        layout->setContentsMargins(0, 0, 0, 0);
+        layout->setSpacing(2);
+        layout->addWidget(m_browseEdit);
+
+        setLayout(layout);
+
+        setFocusProxy(m_browseEdit->lineEdit());
+        setFocusPolicy(m_browseEdit->lineEdit()->focusPolicy());
+    };
+
+    void PropertyFilePathCtrl::SetFilePath(AZ::IO::Path filePath)
+    {
+        m_currentFilePath = filePath;
+        m_browseEdit->setText(filePath.c_str());
+
+        Q_EMIT FilePathChanged();
+    }
+
+    AZ::IO::Path PropertyFilePathCtrl::GetFilePath() const
+    {
+        return m_currentFilePath;
+    }
+
+    void PropertyFilePathCtrl::SetFilter(const QString& filter)
+    {
+        m_filter = filter;
+    }
+
+    void PropertyFilePathCtrl::SetProductExtension(const AZStd::string& extension)
+    {
+        m_productExtension = extension;
+    }
+
+    void PropertyFilePathCtrl::SetDefaultFileName(const AZStd::string& fileName)
+    {
+        m_defaultFileName = fileName;
+    }
+
+    void PropertyFilePathCtrl::OnOpenButtonClicked()
```

Comment:
The guide comments here are super helpful :+1:

suggestion: this is quite a long function, it might be worth splitting out some parts of it and where possible testing certain behaviors in isolation. This makes the testing simpler and helps validate/specify the behavior and avoid any regressions if this code is modified in future (could be a follow-up PR but might be helpful)

Reply:
Good call, I will break it up now. Yeah it was originally a small function but then I kept finding edge cases :)

Reply:
Awesome thanks! 😃